### PR TITLE
Import Remote Zarr Datasets

### LINKED ISSUE #6272: Allow import of remote Zarr datasets
We need a UI to import remote Zarr datasets. Users should be able to enter URLs to their datasets. I imaging that there is a backend route that takes the URL and classifies what kind of data is present, e.g.:
- webKnossos-style dataset (with datasource-properties.json) -> Dataset
- OME-Zarr layer (has .zattrs with OME-NGFF metadata) -> Layer
- Multi-layer OME-Zarr (has .zgroup and OME-layers as sub paths) -> Dataset
- Zarr array (just a zarr array with .zarray) -> Layer

In case of a detected dataset, the user could import that right away. In case of layers, the user may want to add additional layers (i.e. same procedure) before final import.

Supported schemes are http(s), s3, gcs. There should be UI for adding credentials specific to each of these schemes (http => basic or bearer auth, s3 => access key id+secret key, gcs => credentials json (already present for neuroglancer import)).

I don't have a strong opinion, how the datasets are then stored. One idea that @fm3 (please comment) mentioned, was to put a folder with datasoure-properties.json on a datastore.

### LINKED ISSUE #6385: When fetching zarr chunks, handle 404 gracefully
When loading a remote zarr dataset with chunks missing, the 404 errors are not properly caught. For data loading, all buckets that touch those chunks turn fully black, and for histogram loading, the exception is passed all the way through to the user.

Instead, the fill value should be inserted for the chunk’s area.

### PR DESCRIPTION
**Pull request by fm3**
Adds the possibility to link remote Zarr datasets into webKnossos. An import view calls an explore route that returns a proposed datasource-properties.json. The user can merge more explored layers into that json and send it to a datastore, where it is then stored. The datastore can then proxy the remote zarr data to webKnossos.

Note that the front-end part of this PR was developed as sub-pr #6383 and is now merged into here.

### TODO Backend
 - [x] construct remote base path using file system provider (s3, https)
 - [x] fetch and parse zarray
 - [x] fetch and parse zattrs
 - [x] extract voxel size from transformations
 - [x] extract mags from transformations
 - [x] add credentials to API, use them
 - [x] try both exploreAsZarray and exploreAsNgff, use the one that succeeds
 - [x] guess dataset and layer names
 - [x] use all mags, not just the first one
 - [x] error handling
 - [x] pass voxel size out to dataset
 - [x] assert that axis order is one we can handle
 - [x] parse voxel size units
 - [x] store credentials with mag paths
 - [x] make layers unique by name
 - [x] report?
 - [x] test s3 with credentials
 - [x] test https with basicauth
 - [x] allow url of zattrs file directly
 - [x] more detailed error messages in the report
 - [x] send report also if nothing works
 - [x] ZarrArray should support inverted axis order
 - [x] pass axis order from datasource-properties.json to ZarrArray
 - [x] test more examples
 - [x] detect segmentation layers (could be done by name (segmentation, labels), how to set largest segment id?)
 - [x] clean up code

### Follow-Up/Stretch Goals Backend:
 - [ ] collect layers via directory listings
 - [ ] also look for wk-style dataset metadata (datasource-properties.json)
 - [ ] support older versions of the NGFF spec
 - [ ] support other file systems than s3,https

### Steps to test (backend only, compare sub-pr #6383 for frontend):
- test backend route with `fetch("http://localhost:9000/api/datasets/exploreRemote", {method: 'POST', headers: {'Content-Type': 'application/json'}, body: '[{\"remoteUri\":\"https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4dense_motta_et_al_demo/segmentation\"}]'})`
- other optional fields are `user` and `password` (used as basicauth in https case and s3 accesskey/secretkey in s3 case) – talk to me for testing this with real examples
- should return json with `dataSource` (may be null) and `report` (list of strings, interpret as lines lof logging)
- other ngff urls to try (from https://idr.github.io/ome-ngff-samples/ selected by having x,y,z and NGFF 0.4):
  - https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0047A/4496763.zarr
  - https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0062A/6001240.zarr
  - https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0044A/4007801.zarr (note that T=0 is all black, so wk won’t show data)
- test that datasets with different axis orders can be rendered if the axis order is correctly stated in datasource-properties.json
- check that the axis order for zarr-streaming-served datasets is correctly reported in their datasource-properties.json

### Issues:
- fixes #6272 
- contributes to #6356 
- contributes to #6120
- contributes to #6119
- fixes #6385

------
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Needs datastore update after deployment
- [x] Ready for review

### CODE REVIEW

**Comment by normanrz** [comment]
Weitere URLs zum testen:
- https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4dense_motta_et_al_demo/color
- https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4dense_motta_et_al_demo/segmentation

**Comment by fm3** [comment]
@jstriebel I’d say the backend is ready for review :) No hurry, though, since with the frontend still being designed, the PR is not blocked by this review. Also note that the requirements for the backend may still change slightly, but my guess is that the backend code will not change a lot anymore

**Review by jstriebel — COMMENTED** [review]
Awesome stuff :rocket: I reviewed the backend, mostly LGTM, I have only some minor nitpicks

File: app/models/binary/ExploreRemoteLayerService.scala [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        index += 1
+        nameCandidate = f"${layer.name}_$index"
+      }
+      namesSetMutable.add(nameCandidate)
+      if (nameCandidate == layer.name) {
+        layer
+      } else
+        layer match {
+          case l: ZarrDataLayer         => l.copy(name = nameCandidate)
+          case l: ZarrSegmentationLayer => l.copy(name = nameCandidate)
```
[the remark this answers is not in the document]
Comment (fm3) [reply app/models/binary/ExploreRemoteLayerService.scala:64]: Unfortunately, yes, due to typing. The concrete class has to be known when calling `.copy`, as traits don’t have that

**Comment by fm3** [comment]
@jstriebel Could you have a look at the backend commits since your last review please? I hope the commit messages are descriptive enough, otherwise I’m also happy to jump on a call for this :)

File: CHANGELOG.unreleased.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ For upgrade instructions, please check the [migration guide](MIGRATIONS.released
 - Added support for 64-bit segmentations (uint64). Note that the actual support only covers IDs up to 2^53 - 1 (matches the JavaScript number type). Also note that JSON mappings are only compatible with segment ids which only use the lower 32 bits. [#6317](https://github.com/scalableminds/webknossos/pull/6317)
 - Added a "duplicate" button for annotations. [#6386](https://github.com/scalableminds/webknossos/pull/6386)
 - Added new filter options for the dataset list api at `api/datasets`: `organizationName: String`, `onlyMyOrganization: Boolean`, `uploaderId: String` [#6377](https://github.com/scalableminds/webknossos/pull/6377)
+- The Add Dataset view now hosts an Add Zarr Dataset tab to explore, configure and import remote Zarr datasets. [#6383](https://github.com/scalableminds/webknossos/pull/6383)
```
Comment (jstriebel) [inline CHANGELOG.unreleased.md:24]:
```suggestion
- The Add Dataset view now hosts an Add Zarr Dataset tab to explore, configure and import remote Zarr datasets. [#6335](https://github.com/scalableminds/webknossos/pull/6335)
```

**Review by jstriebel — APPROVED** [review]
The backend changes LGTM :+1:

**Pull request merged**